src/wallet/rpcnames.cpp

```diff
@@ -483,11 +523,81 @@ name_firstupdate ()
   if (name.size () > MAX_NAME_LENGTH)
     throw JSONRPCError (RPC_INVALID_PARAMETER, "the name is too long");
 
-  const valtype rand = ParseHexV (request.params[1], "rand");
-  if (rand.size () > 20)
-    throw JSONRPCError (RPC_INVALID_PARAMETER, "invalid rand value");
+  const bool fixedRand = request.params.size () >= 2 && !request.params[1].isNull ();
+  const bool fixedTxid = request.params.size () >= 3 && !request.params[2].isNull ();
+  valtype rand(20);
+  if (fixedRand)
+    {
+      rand = ParseHexV (request.params[1], "rand");
+      if (rand.size () > 20)
+        throw JSONRPCError (RPC_INVALID_PARAMETER, "invalid rand value");
+    }
+
+  uint256 prevTxid = uint256::ZERO; // if it can't find a txid, force an error
+  if (fixedTxid)
+    {
+      prevTxid = ParseHashV (request.params[2], "txid");
+    }
+  else
+    {
+      // Code slightly duplicates name_scan, but not enough to be able to refactor.
+      /* Make sure the results are valid at least up to the most recent block
+         the user could have gotten from another RPC command prior to now.  */
+      pwallet->BlockUntilSyncedToCurrentChain ();
+
+      LOCK (cs_main);
+
+      for (const auto& item : pwallet->mapWallet)
+        {
+          const CWalletTx& tx = item.second;
+          if (!tx.tx->IsNamecoin ())
+            continue;
+
+          CScript output;
+          CNameScript nameOp;
+          int nOut = -1;
+          for (unsigned i = 0; i < tx.tx->vout.size (); ++i)
+            {
+              CScript curOutput = tx.tx->vout[i].scriptPubKey;
+              const CNameScript cur(curOutput);
+              if (!cur.isNameOp ())
+                continue;
+              if (cur.getNameOp () != OP_NAME_NEW)
+                continue;
+              if (nOut != -1) {
+                LogPrintf ("ERROR: wallet contains tx with multiple"
+                           " name outputs");
+                continue;
+              }
+              nameOp = cur;
+              nOut = i;
+              output = curOutput;
+            }
+
+          if (nOut == -1)
+            continue; // no name outputs found
 
-  const uint256 prevTxid = ParseHashV (request.params[2], "txid");
+          const int depth = tx.GetDepthInMainChain ();
+          if (depth <= 0)
+            continue; // not confirmed
```

Comment:
Why is this relevant?  I think we can just remove this check and not care about whether or not the transaction is confirmed at this point.

Reply:
Well, we can't do a firstupdate on something in the mempool anyway.

Reply:
True, but if it doesn't matter, I think we should go with the simpler thing not the more complex.  In other words, if there is no specific reason why we *need* the pending check, I would just remove it to simplify the code.  Or if there is a specific reason why we really need the check here, it should be explained in a comment.

Reply:
All right, yeah, I was cargo-culting `name_list` there...

You might want to check >12, but it seems better that it should give an "immature name" error than a generic "scan failed". On the other hand, if you have one mature and one immature, it also seems good that it would choose the immature.

Could this introduce a DoS vulnerability? Say I'm trying to do `name_firstupdate` on a name, but someone spams me with immature name_new for the same name. But that's ridiculous, and if it came to that I could just fill in the TXID manually, right.

So I think we'll do fine skipping it without anything else to replace it.